File: code/ledgerscope-accounts/src/Ledgerscope.Accounts.Domain/Cases/Case.cs
namespace Ledgerscope.Accounts.Domain.Cases;

/// <summary>
/// Investigation aggregate: a container of the addresses, assignments, and
/// notes that make up one case. The graph, risk score, and tracing always
/// live in the Rust engine — a case only records which addresses belong to
/// the investigation plus the metadata around them.
/// </summary>
public sealed class Case {
    private readonly List<CaseAddress> addresses = [];
    private readonly List<CaseAssignment> assignments = [];
    private readonly List<CaseNote> notes = [];

    private Case() {
    }

    public Case(
        Guid id, String title, String description, CasePriority priority,
        Guid organizationId, Guid createdBy, DateTimeOffset createdAt) {
        Id = id;
        Title = title;
        Description = description;
        Priority = priority;
        OrganizationId = organizationId;
        CreatedBy = createdBy;
        CreatedAt = createdAt;
        Status = CaseStatus.Open;
    }

    public Guid Id { get; private set; }
    public String Title { get; private set; } = String.Empty;
    public String Description { get; private set; } = String.Empty;
    public CaseStatus Status { get; private set; }
    public CasePriority Priority { get; private set; }
    public Guid OrganizationId { get; private set; }
    public Guid CreatedBy { get; private set; }
    public DateTimeOffset CreatedAt { get; private set; }
    public DateTimeOffset? ClosedAt { get; private set; }

    public IReadOnlyCollection<CaseAddress> Addresses => addresses;
    public IReadOnlyCollection<CaseAssignment> Assignments => assignments;
    public IReadOnlyCollection<CaseNote> Notes => notes;

    public void UpdateDetails(String title, String description, CasePriority priority) {
        Title = title;
        Description = description;
        Priority = priority;
    }

    public void Close(DateTimeOffset when) {
        Status = CaseStatus.Closed;
        ClosedAt = when;
    }

    public void Reopen() {
        Status = CaseStatus.Reopened;
        ClosedAt = null;
    }

    public CaseAssignment Assign(Guid userId, String roleName, Guid assignedBy, DateTimeOffset when) {
        var existing = assignments.FirstOrDefault(assignment => assignment.UserId == userId);
        if (existing is not null) {
            existing.ChangeRole(roleName, assignedBy, when);
            return existing;
        }

        var created = new CaseAssignment(Guid.NewGuid(), Id, userId, roleName, assignedBy, when);
        assignments.Add(created);
        return created;
    }

    public void Unassign(Guid userId) {
        assignments.RemoveAll(assignment => assignment.UserId == userId);
    }

    public CaseAddress AddAddress(
        String address, Int32 chainId, Guid addedBy, DateTimeOffset when, String? note) {
        var existing = addresses.FirstOrDefault(
            item => item.Address == address && item.ChainId == chainId);
        if (existing is not null) {
            return existing;
        }

        var created = new CaseAddress(Guid.NewGuid(), Id, address, chainId, addedBy, when, note);
        addresses.Add(created);
        return created;
    }

    public void RemoveAddress(String address, Int32 chainId) {
        addresses.RemoveAll(item => item.Address == address && item.ChainId == chainId);
    }

    public CaseNote AddNote(Guid authorId, String text, DateTimeOffset when) {
        var created = new CaseNote(Guid.NewGuid(), Id, authorId, text, when);
        notes.Add(created);
        return created;
    }
}

File: code/frontend/src/entities/case/model/case.ts
export type CaseStatus = 'Open' | 'InProgress' | 'Closed' | 'Reopened'
export type CasePriority = 'Low' | 'Medium' | 'High' | 'Critical'

export const CASE_PRIORITIES: CasePriority[] = ['Low', 'Medium', 'High', 'Critical']
export const CASE_STATUSES: CaseStatus[] = ['Open', 'InProgress', 'Closed', 'Reopened']

export type CaseSummary = {
  id: string
  title: string
  status: CaseStatus
  priority: CasePriority
  organizationId: string
  createdAt: string
}

export type CaseAddress = {
  address: string
  chainId: number
  addedBy: string
  addedAt: string
  note: string | null
}

export type CaseAssignment = {
  userId: string
  roleName: string
  assignedAt: string
}

export type CaseNote = {
  id: string
  authorId: string
  text: string
  createdAt: string
}

export type CaseDetail = {
  id: string
  title: string
  description: string
  status: CaseStatus
  priority: CasePriority
  organizationId: string
  createdBy: string
  createdAt: string
  closedAt: string | null
  addresses: CaseAddress[]
  assignments: CaseAssignment[]
  notes: CaseNote[]
}

export const STATUS_LABEL: Record<CaseStatus, string> = {
  Open: 'Open',
  InProgress: 'In progress',
  Closed: 'Closed',
  Reopened: 'Reopened',
}

/** Tailwind classes for a status pill. */
export function statusClasses(status: CaseStatus): string {
  switch (status) {
    case 'Open':
      return 'bg-accent/15 text-accent border-accent/30'
    case 'InProgress':
      return 'bg-primary/15 text-primary border-primary/30'
    case 'Reopened':
      return 'bg-warning/15 text-warning border-warning/30'
    case 'Closed':
      return 'bg-muted text-muted-foreground border-border'
  }
}

/** Tailwind classes for a priority pill. */
export function priorityClasses(priority: CasePriority): string {
  switch (priority) {
    case 'Low':
      return 'bg-muted text-muted-foreground border-border'
    case 'Medium':
      return 'bg-accent/15 text-accent border-accent/30'
    case 'High':
      return 'bg-warning/15 text-warning border-warning/30'
    case 'Critical':
      return 'bg-destructive/15 text-destructive border-destructive/30'
  }
}
